tst/testinstall/kernel/hpc/threadapi.tst

```diff
@@ -0,0 +1,397 @@
+gap> MultiSendChannel(fail, fail);
+Error, MultiSendChannel: <channel> must be a channel (not the value 'fail')
+gap> MultiTransmitChannel(fail, fail);
+Error, MultiTransmitChannel: <channel> must be a channel (not the value 'fail'\
+)
+gap> TryMultiSendChannel(fail, fail);
+Error, TryMultiSendChannel: <channel> must be a channel (not the value 'fail')
+gap> TryMultiTransmitChannel(fail, fail);
+Error, TryMultiTransmitChannel: <channel> must be a channel (not the value 'fa\
+il')
+gap> TrySendChannel(fail, fail);
+Error, TrySendChannel: <channel> must be a channel (not the value 'fail')
+gap> TryTransmitChannel(fail, fail);
+Error, TryTransmitChannel: <channel> must be a channel (not the value 'fail')
+gap> ReceiveAnyChannel(fail);
+Error, ReceiveAnyChannel: Argument list must be channels
+gap> ReceiveAnyChannel([]);
+Error, ReceiveAnyChannel: Argument list must be channels
+gap> ReceiveAnyChannelWithIndex(fail);
+Error, ReceiveAnyChannelWithIndex: Argument list must be channels
+gap> ReceiveAnyChannelWithIndex([]);
+Error, ReceiveAnyChannelWithIndex: Argument list must be channels
+gap> MultiReceiveChannel(fail, fail);
+Error, MultiReceiveChannel: <channel> must be a channel (not the value 'fail')
+gap> InspectChannel(fail);
+Error, InspectChannel: <channel> must be a channel (not the value 'fail')
+gap> TryReceiveChannel(fail, fail);
+Error, TryReceiveChannel: <channel> must be a channel (not the value 'fail')
+
+#
+# semaphores
+#
+gap> CreateSemaphore(fail);
+Error, CreateSemaphore: Argument must be initial count
+gap> CreateSemaphore(-1);
+Error, CreateSemaphore: Initial count must be non-negative
+gap> CreateSemaphore(fail, fail);
+Error, CreateSemaphore: Function takes up to two arguments
+gap> SignalSemaphore(fail);
+Error, SignalSemaphore: <semaphore> must be a semaphore (not the value 'fail')
+gap> WaitSemaphore(fail);
+Error, WaitSemaphore: <semaphore> must be a semaphore (not the value 'fail')
+gap> TryWaitSemaphore(fail);
+Error, TryWaitSemaphore: <semaphore> must be a semaphore (not the value 'fail'\
+)
+
+#
+gap> sem := CreateSemaphore(0);
+<semaphore with count = 0>
+gap> TryWaitSemaphore(sem);
+false
+gap> SignalSemaphore(sem);
+gap> sem;
+<semaphore with count = 1>
+gap> TryWaitSemaphore(sem);
+true
+gap> sem;
+<semaphore with count = 0>
+
+#
+# barriers
+#
+gap> b := CreateBarrier();
+<barrier with 0 of 0 threads arrived>
+gap> StartBarrier(fail, fail);
+Error, StartBarrier: <barrier> must be a barrier (not the value 'fail')
+gap> StartBarrier(b, fail);
+Error, StartBarrier: <count> must be a small integer (not the value 'fail')
+gap> WaitBarrier(fail);
+Error, WaitBarrier: <barrier> must be a barrier (not the value 'fail')
+
+#
+gap> StartBarrier(b, 2);
+gap> b;
+<barrier with 0 of 2 threads arrived>
+gap> tmp:=false;;
+gap> t:=CreateThread(function() WaitBarrier(b); repeat MicroSleep(10); until tmp; end);;
+gap> MicroSleep(50);
+gap> b;
+<barrier with 1 of 2 threads arrived>
```

Comment:
@rbehrends Either I am misunderstanding something (which wouldn't be a surprise, given that barriers in HPC-GAP seem to be completely undocumented), or there is a bug: but it seems that sometimes, the first thread created above does not "arrive" so we get "0 of 2" printed above. Even if I change the `MicroSleep` above to wait for a full second.

Reply:
HPC-GAP barrier semantics follow the semantics POSIX barriers fairly precisely. I.e. after calling `StartBarrier(b, n)`, `WaitBarrier(b)` will block until it has been called exactly `n` times on `b` (this also necessitates that there will be `n` distinct threads doing that).

There is an issue with printing barrier objects in that the numbers may be wrong. I've added a `hpcgap-barriers` branch on https://github.com/rbehrends/gap that should fix the printing issue.

Changing the second `MicroSleep()` above to be longer should properly report a change in state, though (I have also not been able to reproduce the issue doing that). Unless you are talking about the `MicroSleep()` inside the thread function, which should typically have no effect on the barrier state at that point.